File: spark/common/secret_sharing/secret_sharing.go
package secretsharing

import (
	"crypto/rand"
	"fmt"
	"math/big"

	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/lightsparkdev/spark/common"
	pb "github.com/lightsparkdev/spark/proto/spark"
)

// Polynomial is a polynomial used for secret sharing.
type Polynomial struct {
	// FieldModulus is the field modulus of the polynomial.
	FieldModulus *big.Int
	// coefficients are the coefficients of the polynomial.
	coefficients []*big.Int
	// Proofs are the proofs of the polynomial.
	Proofs [][]byte
}

// LagrangeInterpolatable is an interface that can be used to interpolate a secret from a set of shares.
type LagrangeInterpolatable interface {
	// GetIndex returns the index of the share.
	GetIndex() *big.Int
	// GetShare returns the share of the secret.
	GetShare() *big.Int
	// GetFieldModulus returns the field modulus of the share.
	GetFieldModulus() *big.Int
	// GetThreshold returns the threshold of the secret.
	GetThreshold() int
}

// SecretShare is a share of a secret.
type SecretShare struct {
	// FieldModulus is the field modulus of the share.
	FieldModulus *big.Int
	// Threshold is the threshold of the secret.
	Threshold int
	// Index is the index of the share.
	Index *big.Int
	// Share is the share of the secret.
	Share *big.Int
}

// GetIndex returns the index of the share.
func (s *SecretShare) GetIndex() *big.Int {
	return s.Index
}

// GetFieldModulus returns the field modulus of the share.
func (s *SecretShare) GetFieldModulus() *big.Int {
	return s.FieldModulus
}

// GetShare returns the share of the secret.
func (s *SecretShare) GetShare() *big.Int {
	return s.Share
}

// IntToScalar converts a big.Int to a secp256k1 ModNScalar.
func IntToScalar(value *big.Int) *secp256k1.ModNScalar {
	var scalar secp256k1.ModNScalar
	scalar.SetByteSlice(value.Bytes())
	return &scalar
}

// GetThreshold returns the threshold of the secret.
func (s *SecretShare) GetThreshold() int {
	return s.Threshold
}

// VerifiableSecretShare is a share of a secret with proofs.
type VerifiableSecretShare struct {
	SecretShare
	Proofs [][]byte
}

// MarshalProto marshals the VerifiableSecretShare to a protobuf message.
func (v *VerifiableSecretShare) MarshalProto() *pb.SecretShare {
	return &pb.SecretShare{
		SecretShare: v.Share.Bytes(),
		Proofs:      v.Proofs,
	}
}

// Evaluate evaluates the polynomial at a given point.
func (p *Polynomial) Evaluate(x *big.Int) *big.Int {
	result := big.NewInt(0)
	for i, coeff := range p.coefficients {
		xPow := new(big.Int).Exp(x, big.NewInt(int64(i)), p.FieldModulus)
		xPow.Mul(xPow, coeff)
		result.Add(result, xPow)
		result.Mod(result, p.FieldModulus)
	}
	return result
}

// fieldDiv divides two numbers in a given field modulus.
func fieldDiv(numerator *big.Int, denominator *big.Int, fieldModulus *big.Int) (*big.Int, error) {
	if denominator.Sign() == 0 {
		return nil, fmt.Errorf("division by zero")
	}

	inverse := new(big.Int).ModInverse(denominator, fieldModulus)
	inverse.Mul(inverse, numerator)
	return inverse.Mod(inverse, fieldModulus), nil
}

// ComputeLagrangeCoefficients computes the Lagrange coefficient for a given index and a set of points.
func ComputeLagrangeCoefficients[T LagrangeInterpolatable](index *big.Int, points []T) (*big.Int, error) {
	numerator := big.NewInt(1)
	denominator := big.NewInt(1)
	fieldModulus := points[0].GetFieldModulus()
	for _, point := range points {
		if point.GetIndex().Cmp(index) == 0 {
			continue
		}
		numerator.Mul(numerator, point.GetIndex())
		value := new(big.Int).Sub(point.GetIndex(), index)
		denominator.Mul(denominator, value)
	}

	return fieldDiv(numerator, denominator, fieldModulus)
}

// generatePolynomialForSecretSharing generates a polynomial for secret sharing.
func generatePolynomialForSecretSharing(fieldModulus *big.Int, secret *big.Int, degree int) (*Polynomial, error) {
	coefficients := make([]*big.Int, degree)
	proofs := make([][]byte, degree)

	coefficients[0] = secret
	var secretScalar secp256k1.ModNScalar
	secretScalar.SetByteSlice(secret.Bytes())
	proofs[0] = secp256k1.NewPrivateKey(&secretScalar).PubKey().SerializeCompressed()
	for i := 1; i < degree; i++ {
		randomInt, err := rand.Int(rand.Reader, fieldModulus)
		if err != nil {
			return nil, err
		}
		coefficients[i] = randomInt
		var coefScalar secp256k1.ModNScalar
		coefScalar.SetByteSlice(randomInt.Bytes())
		proofs[i] = secp256k1.NewPrivateKey(&coefScalar).PubKey().SerializeCompressed()
	}
	return &Polynomial{
		FieldModulus: fieldModulus,
		coefficients: coefficients,
		Proofs:       proofs,
	}, nil
}

// SplitSecret splits a secret into a set of shares.
func SplitSecret(secret *big.Int, fieldModulus *big.Int, threshold int, numberOfShares int) ([]*SecretShare, error) {
	polynomial, err := generatePolynomialForSecretSharing(fieldModulus, secret, threshold)
	if err != nil {
		return nil, err
	}

	shares := make([]*SecretShare, 0)
	for i := 1; i <= numberOfShares; i++ {
		share := polynomial.Evaluate(big.NewInt(int64(i)))
		shares = append(shares, &SecretShare{
			FieldModulus: fieldModulus,
			Threshold:    threshold,
			Index:        big.NewInt(int64(i)),
			Share:        share,
		})
	}
	return shares, nil
}

// SplitSecretWithProofs splits a secret into a set of shares with proofs.
func SplitSecretWithProofs(secret *big.Int, fieldModulus *big.Int, threshold int, numberOfShares int) ([]*VerifiableSecretShare, error) {
	polynomial, err := generatePolynomialForSecretSharing(fieldModulus, secret, threshold-1)
	if err != nil {
		return nil, err
	}

	shares := make([]*VerifiableSecretShare, 0)
	for i := 1; i <= numberOfShares; i++ {
		share := polynomial.Evaluate(big.NewInt(int64(i)))
		shares = append(shares, &VerifiableSecretShare{
			SecretShare: SecretShare{
				FieldModulus: fieldModulus,
				Threshold:    threshold,
				Index:        big.NewInt(int64(i)),
				Share:        share,
			},
			Proofs: polynomial.Proofs,
		})
	}
	return shares, nil
}

// RecoverSecret recovers a secret from a set of shares.
func RecoverSecret[T LagrangeInterpolatable](shares []T) (*big.Int, error) {
	if len(shares) < shares[0].GetThreshold() {
		return nil, fmt.Errorf("not enough shares to recover secret")
	}

	result := big.NewInt(0)
	for _, share := range shares {
		coeff, err := ComputeLagrangeCoefficients(share.GetIndex(), shares)
		if err != nil {
			return nil, err
		}
		item := new(big.Int).Mul(share.GetShare(), coeff)
		item.Mod(item, shares[0].GetFieldModulus())
		result.Add(result, item)
		result.Mod(result, shares[0].GetFieldModulus())
	}

	return result, nil
}

// ValidateShare validates a share of a secret.
func ValidateShare(share *VerifiableSecretShare) error {
	targetPubkey := secp256k1.NewPrivateKey(IntToScalar(share.Share)).PubKey()
	resultPubkey, err := secp256k1.ParsePubKey(share.Proofs[0])
	if err != nil {
		return err
	}
	for i, proof := range share.Proofs {
		if i == 0 {
			continue
		}
		pubkey, err := secp256k1.ParsePubKey(proof)
		if err != nil {
			return err
		}

		value := new(big.Int).Exp(share.Index, big.NewInt(int64(i)), share.FieldModulus)
		curve := secp256k1.S256()
		resX, resY := curve.ScalarMult(pubkey.X(), pubkey.Y(), value.Bytes())
		resPubkey := common.PublicKeyFromInts(resX, resY)

		resultPubkey = common.AddPublicKeysRaw(resultPubkey, resPubkey)
	}

	if resultPubkey.X().Cmp(targetPubkey.X()) != 0 || resultPubkey.Y().Cmp(targetPubkey.Y()) != 0 {
		return fmt.Errorf("share is not valid")
	}

	return nil
}

File: sdks/js/packages/spark-sdk/src/utils/secret-sharing.ts
import { bytesToHex, equalBytes } from "@noble/curves/abstract/utils";
import { secp256k1 } from "@noble/curves/secp256k1";
import { getCrypto } from "./crypto.js";
import { ValidationError } from "../errors/index.js";

const crypto = getCrypto();

type Polynomial = {
  fieldModulus: bigint;
  coefficients: bigint[];
  proofs: Uint8Array[];
};

type SecretShare = {
  fieldModulus: bigint;
  threshold: number;
  index: bigint;
  share: bigint;
};

export type VerifiableSecretShare = SecretShare & {
  proofs: Uint8Array[];
};

export function getRandomBigInt(max: bigint): bigint {
  const byteLength = (max.toString(2).length + 7) >> 3;
  const maxBigInt = max;

  const mask = (1n << BigInt(max.toString(2).length)) - 1n;
  while (true) {
    const randBytes = crypto.getRandomValues(new Uint8Array(byteLength + 1));

    const randValue = BigInt("0x" + bytesToHex(randBytes)) & mask;

    if (randValue < maxBigInt) {
      return randValue;
    }
  }
}

// Modular inverse using extended euclidean algorithm
export function modInverse(a: bigint, m: bigint): bigint {
  // Handle negative numbers by making them positive
  a = ((a % m) + m) % m;

  let [old_r, r] = [a, m];
  let [old_s, s] = [1n, 0n];
  let [old_t, t] = [0n, 1n];

  while (r !== 0n) {
    const quotient = old_r / r;
    [old_r, r] = [r, old_r - quotient * r];
    [old_s, s] = [s, old_s - quotient * s];
    [old_t, t] = [t, old_t - quotient * t];
  }

  if (old_r !== 1n) {
    throw new ValidationError("Modular inverse does not exist", {
      field: "modInverse",
      value: `a: ${a}, m: ${m}`,
      expected: "a and m must be coprime",
    });
  }

  return ((old_s % m) + m) % m;
}

// Evaluates a polynomial at a given point
export function evaluatePolynomial(polynomial: Polynomial, x: bigint): bigint {
  let result = 0n;
  for (let i = 0; i < polynomial.coefficients.length; i++) {
    const coeff = polynomial.coefficients[i];
    if (!coeff) {
      throw new ValidationError("Coefficient is undefined", {
        field: "coefficient",
        value: "undefined",
        expected: "A valid bigint coefficient",
      });
    }

    const xPow = x ** BigInt(i) % polynomial.fieldModulus;

    result = (result + xPow * coeff) % polynomial.fieldModulus;
  }
  return result;
}

// Divides two numbers in a given field modulus
export function fieldDiv(
  numerator: bigint,
  denominator: bigint,
  fieldModulus: bigint,
): bigint {
  if (denominator === 0n) {
    throw new ValidationError("Division by zero", {
      field: "denominator",
      value: "0",
      expected: "Non-zero value",
    });
  }

  const inverse = modInverse(denominator, fieldModulus);
  return (numerator * inverse) % fieldModulus;
}

// Computes the Lagrange coefficient for a given index and a set of points
export function computerLagrangeCoefficients(
  index: bigint,
  points: SecretShare[],
) {
  let numerator = 1n;
  let denominator = 1n;
  let fieldModulus = points[0]?.fieldModulus;
  if (!fieldModulus) {
    throw new ValidationError("Field modulus is undefined", {
      field: "fieldModulus",
      value: "undefined",
      expected: "A valid field modulus",
    });
  }

  for (const point of points) {
    if (point.index === index) {
      continue;
    }
    numerator = numerator * point.index;
    const value = point.index - index;
    denominator = denominator * value;
  }

  return fieldDiv(numerator, denominator, fieldModulus);
}

// Generates a polynomial for secret sharing
export function generatePolynomialForSecretSharing(
  fieldModulus: bigint,
  secret: bigint,
  degree: number,
): Polynomial {
  const coefficients: bigint[] = new Array(degree);
  const proofs: Uint8Array[] = new Array(degree);

  coefficients[0] = secret;
  proofs[0] = secp256k1.ProjectivePoint.fromPrivateKey(secret).toRawBytes(true);

  for (let i = 1; i < degree; i++) {
    const coefficient = getRandomBigInt(fieldModulus);
    coefficients[i] = coefficient;
    proofs[i] =
      secp256k1.ProjectivePoint.fromPrivateKey(coefficient).toRawBytes(true);
  }
  return {
    fieldModulus,
    coefficients,
    proofs: proofs,
  };
}

// Splits a secret into a list of shares
export function splitSecret(
  fieldModulus: bigint,
  secret: bigint,
  threshold: number,
  numberOfShares: number,
) {
  const polynomial = generatePolynomialForSecretSharing(
    fieldModulus,
    secret,
    threshold,
  );

  const shares: SecretShare[] = [];
  for (let i = 1; i <= numberOfShares; i++) {
    const share = evaluatePolynomial(polynomial, BigInt(i));
    shares.push({
      fieldModulus,
      threshold,
      index: BigInt(i),
      share,
    });
  }

  return shares;
}

// Splits a secret into a list of shares with proofs
export function splitSecretWithProofs(
  secret: bigint,
  fieldModulus: bigint,
  threshold: number,
  numberOfShares: number,
) {
  const polynomial = generatePolynomialForSecretSharing(
    fieldModulus,
    secret,
    threshold - 1,
  );

  const shares: VerifiableSecretShare[] = [];
  for (let i = 1; i <= numberOfShares; i++) {
    const share = evaluatePolynomial(polynomial, BigInt(i));
    shares.push({
      fieldModulus,
      threshold,
      index: BigInt(i),
      share,
      proofs: polynomial.proofs,
    });
  }

  return shares;
}

// Recovers a secret from a list of shares
export function recoverSecret(shares: VerifiableSecretShare[]) {
  if (shares.length === 0) return 0n;

  const threshold = shares[0]?.threshold;
  const fieldModulus = shares[0]?.fieldModulus;

  if (!threshold || !fieldModulus) {
    throw new ValidationError("Shares are not valid", {
      field: "shares",
      value: "Missing threshold or fieldModulus",
      expected: "Valid shares with threshold and fieldModulus",
    });
  }

  if (shares.length < threshold) {
    throw new ValidationError("Not enough shares to recover secret", {
      field: "shares",
      value: shares.length,
      expected: `At least ${threshold} shares`,
    });
  }

  let result = 0n;
  for (const share of shares) {
    const coeff = computerLagrangeCoefficients(share.index, shares);
    const item = (share.share * coeff) % fieldModulus;

    result = (result + item) % fieldModulus;
  }

  return result;
}

// Validates a share of a secret
export function validateShare(share: VerifiableSecretShare) {
  const targetPubkey = secp256k1.ProjectivePoint.fromPrivateKey(
    share.share,
  ).toRawBytes(true);

  let resultPubkey = share.proofs[0];
  if (!resultPubkey) {
    throw new ValidationError("Result pubkey is not valid", {
      field: "resultPubkey",
      value: "null",
      expected: "Valid public key bytes",
    });
  }

  for (let i = 1; i < share.proofs.length; i++) {
    const pubkey = share.proofs[i];
    if (!pubkey) {
      throw new ValidationError("Pubkey is not valid", {
        field: "pubkey",
        value: "null",
        expected: "Valid public key bytes",
      });
    }
    const value = share.index ** BigInt(i) % share.fieldModulus;

    const scaledPoint =
      secp256k1.ProjectivePoint.fromHex(pubkey).multiply(value);
    resultPubkey = secp256k1.ProjectivePoint.fromHex(resultPubkey)
      .add(scaledPoint)
      .toRawBytes(true);
  }

  if (!equalBytes(resultPubkey, targetPubkey)) {
    throw new ValidationError("Share is not valid", {
      field: "share",
      value: "Invalid proof",
      expected: "Valid share with matching proofs",
    });
  }
}

// Converts a bigint to a private key since imported package doesn't support bigint
export function bigIntToPrivateKey(value: bigint): Uint8Array {
  const hex = value.toString(16).padStart(64, "0");

  const bytes = new Uint8Array(32);
  for (let i = 0; i < 32; i++) {
    bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }

  return bytes;
}
